Add new crc32 tests and optimized crc32 for POWER

mscastanho opened the pull request:
Hi,

This PR adds an optimized version of the crc32 function for POWER processors. This is a port of madler/zlib#478 to zlib-ng. The algorithm is mostly intact, as originally developed by @racardoso and added to the original zlib by @grooverdan.

While preparing this PR, I also found some issues in the way `$ARCH` was detected on the CI using CMake, so I added a fix.

This PR also adds new crc32 tests that can be used by all targets.

I'm marking this as a Draft PR for a few reasons:
- I had to slightly change the crc32_power8 function API to match zlib-ng's, but there may be still some small changes to be made. So I'd like some feedback on that (@racardoso).
- I'm unsure about the correct line to setup the pointer for `crc32_power8` on `functable.c` as it's not quite clear to me what `if (sizeof(void *) == sizeof(ptrdiff_t))` is checking.
- I kept the addition of the new tests + optimized function in the same commit as it was in the original PR. But I can split those into two if you prefer.

To measure the performance improvement, we used [Chromium's zlib_bench.cc](https://cs.chromium.org/chromium/src/third_party/zlib/contrib/bench/) with input files from [jsnell/zlib-bench](https://github.com/jsnell/zlib-bench/tree/master/corpus).

The results below show compression and decompression throughput in MB/s using gzip wrapper, for all compression levels:

```
======== File 'executable' ======== 

 Wrapper 'gzip'

          |   compression throughput (MB)    |  decompression throughput (MB)   |
  | level | default | crc32-power |     gain | default | crc32-power |     gain |
  |     1 |    90.0 |       101.7 |  +13.00% |   187.1 |       254.7 |  +36.13% |
  |     2 |    59.9 |        65.2 |   +8.85% |   169.3 |       228.4 |  +34.91% |
  |     3 |    50.6 |        54.7 |   +8.10% |   174.3 |       237.4 |  +36.20% |
  |     4 |    46.0 |        49.3 |   +7.17% |   178.5 |       244.6 |  +37.03% |
  |     5 |    42.3 |        44.6 |   +5.44% |   180.0 |       247.4 |  +37.44% |
  |     6 |    36.2 |        37.9 |   +4.70% |   181.0 |       249.3 |  +37.73% |
  |     7 |    21.3 |        22.0 |   +3.29% |   182.5 |       251.9 |  +38.03% |
  |     8 |    14.1 |        14.4 |   +2.13% |   182.9 |       252.4 |  +38.00% |
  |     9 |     9.2 |         9.4 |   +2.17% |   183.1 |       252.8 |  +38.07% |

======== File 'pngpixels' ======== 

 Wrapper 'gzip'

          |   compression throughput (MB)    |  decompression throughput (MB)   |
  | level | default | crc32-power |     gain | default | crc32-power |     gain |
  |     1 |   128.5 |       159.6 |  +24.20% |   253.7 |       398.3 |  +57.00% |
  |     2 |   111.3 |       132.0 |  +18.60% |   265.8 |       430.4 |  +61.93% |
  |     3 |    81.1 |        92.3 |  +13.81% |   288.3 |       495.1 |  +71.73% |
  |     4 |    71.8 |        80.8 |  +12.53% |   286.4 |       490.3 |  +71.19% |
  |     5 |    54.9 |        59.3 |   +8.01% |   306.4 |       551.1 |  +79.86% |
  |     6 |    29.9 |        31.3 |   +4.68% |   327.0 |       617.9 |  +88.96% |
  |     7 |    16.7 |        17.1 |   +2.40% |   326.8 |       613.5 |  +87.73% |
  |     8 |     5.3 |         5.3 |   +0.00% |   331.3 |       631.0 |  +90.46% |
  |     9 |     2.5 |         2.5 |   +0.00% |   332.1 |       635.4 |  +91.33% |

======== File 'jpeg' ======== 

 Wrapper 'gzip'

          |   compression throughput (MB)    |  decompression throughput (MB)   |
  | level | default | crc32-power |     gain | default | crc32-power |     gain |
  |     1 |    56.7 |        62.0 |   +9.35% |   155.2 |       199.7 |  +28.67% |
  |     2 |    26.1 |        27.2 |   +4.21% |   527.7 |      2276.4 | +331.38% |
  |     3 |    26.1 |        27.2 |   +4.21% |   527.7 |      2270.9 | +330.34% |
  |     4 |    22.1 |        22.9 |   +3.62% |   527.8 |      2275.2 | +331.07% |
  |     5 |    22.1 |        22.9 |   +3.62% |   527.6 |      2269.5 | +330.16% |
  |     6 |    22.1 |        22.9 |   +3.62% |   527.8 |      2275.9 | +331.21% |
  |     7 |    25.6 |        25.6 |   +0.00% |   527.9 |      2278.2 | +331.56% |
  |     8 |    25.6 |        26.7 |   +4.30% |   527.5 |      2278.2 | +331.89% |
  |     9 |    25.6 |        25.5 |   -0.39% |   527.8 |      2277.3 | +331.47% |

======== File 'html' ======== 

 Wrapper 'gzip'

          |   compression throughput (MB)    |  decompression throughput (MB)   |
  | level | default | crc32-power |     gain | default | crc32-power |     gain |
  |     1 |    81.0 |        91.4 |  +12.84% |   164.2 |       215.1 |  +31.00% |
  |     2 |    61.9 |        67.9 |   +9.69% |   191.3 |       261.6 |  +36.75% |
  |     3 |    47.7 |        51.2 |   +7.34% |   199.5 |       277.4 |  +39.05% |
  |     4 |    45.9 |        49.0 |   +6.75% |   203.6 |       286.0 |  +40.47% |
  |     5 |    41.6 |        43.7 |   +5.05% |   205.5 |       289.1 |  +40.68% |
  |     6 |    34.6 |        36.3 |   +4.91% |   206.9 |       291.7 |  +40.99% |
  |     7 |    22.3 |        23.0 |   +3.14% |   202.4 |       283.5 |  +40.07% |
  |     8 |    15.7 |        16.0 |   +1.91% |   202.3 |       283.6 |  +40.19% |
  |     9 |    15.4 |        15.7 |   +1.95% |   202.1 |       283.5 |  +40.28% |
```

nmoinvaz commented on test/crc32_test.c:
```diff
@@ -0,0 +1,198 @@
+
+typedef struct {
+    int line;
+    uLong crc;
```
Use `uint32_t` instead of `uLong`.

mscastanho replied:
Done.

nmoinvaz commented on test/crc32_test.c:
```diff
@@ -0,0 +1,198 @@
+    char* buf;
+    int len;
+    uLong expect;
+} crc32_test;
```
Use standard fixed int types in this struct.

mscastanho replied:
I actually ifdef'd this struct's definition to match the different crc32 declarations for the compat and zlib-ng APIs. When building the default API, it will use fixed int types.
This way we was use the proper argument types and avoid potential problems with "hidden" type casting.

nmoinvaz commented on configure:
```diff
@@ -1392,6 +1392,13 @@ case "${ARCH}" in
+                    powerpc64*)
+                        ARCH_STATIC_OBJS="${ARCH_STATIC_OBJS} crc32_power8.o"
+                        ARCH_SHARED_OBJS="${ARCH_SHARED_OBJS} crc32_power8.lo"
+                    ;;
```
Suggest new `-DPOWER8_CRC32`.

mscastanho replied:
Done.

nmoinvaz commented on functable.c:
```diff
@@ -59,6 +59,8 @@ ZLIB_INTERNAL uint32_t crc32_generic(uint32_t, const unsigned char *, uint64_t);
 
 #ifdef __ARM_FEATURE_CRC32
 extern uint32_t crc32_acle(uint32_t, const unsigned char *, uint64_t);
+#elif defined(POWER8) && (defined(__powerpc64__) || defined(__PPC64__))
```
Use `#elif defined(POWER8_CRC32)` instead.

nmoinvaz commented on functable.c:
```diff
@@ -223,6 +225,7 @@ ZLIB_INTERNAL uint32_t crc32_stub(uint32_t crc, const unsigned char *buf, uint64
            "crc32_z takes size_t but internally we have a uint64_t len");
     /* return a function pointer for optimized arches here after a capability test */
 
+    functable.crc32 = &crc32_generic;
```
Good change.

nmoinvaz commented:
This code is GPL or Apache? Can you get it relicensed under zlib license?

nmoinvaz commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,671 @@
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of either:
+ *
+ *  a) the GNU General Public License as published by the Free Software
```
GPL license..

mscastanho commented:
@nmoinvaz today the code is dual licensed (GPL / Apache). I'm checking to see what we can do in that regard.

mtl1979 commented:
All code should be licensed using zlib license, this avoids any viral effects of licenses like GPL.

mscastanho commented:
Incorporated suggestions from @nmoinvaz and rebased against latest `develop` branch.

nmoinvaz commented:
@mscastanho were you able to resolve the licensing issues?

mscastanho commented:
@nmoinvaz No updates on this so far =/ I believe this won't make to the first release.

Dead2 commented:
This needs a rebase.
Perhaps the crc32 tests could be split out into a separate PR? Looks like there is no licensing problem with that part, and it would be useful for all platforms.

mscastanho commented:
@Dead2 Agreed. I'll split the tests into a separate PR and submit it.
Regarding the license issues for the optimization itself, I haven't been able to sort it out yet.

mscastanho committed:
Standardize crc32_stub

Reorganize statements inside crc32_stub() to match more closely the format
used for other function stubs in functable.c.

mscastanho committed:
Add new crc32 unit test

mscastanho committed:
Add optimized crc32 for POWER8 and later processors

This commit adds an optimized version of the crc32 function based
on crc32-vpmsum from https://github.com/antonblanchard/crc32-vpmsum/ .
The code has been relicensed to the zlib license.

This is the C implementation created by Rogerio Alves <rogealve@br.ibm.com>

It makes use of vector instructions to speed up CRC32 algorithm. Decompression
times were improved by +30% on tests.

Based on Daniel Black's work for the original zlib (madler/zlib#478).

mscastanho commented:
@Dead2 

- PR rebased against latest develop.
- I was also able to relicense the code to zlib license.
- Addition of the tests are in a separa commit now
- Removed one commit that was not needed anymore

I believe everything is now ready for review.

nmoinvaz commented on test/crc32_test.c:
```diff
@@ -0,0 +1,208 @@
+typedef unsigned long crc32_type;
+typedef unsigned char buf_type;
+typedef unsigned int len_type;
+#  define CRC_FMT "%08lX"
```
I wonder if this `CRC_FMT` should be moved into zlib-ng.h. @Dead2 looks like z_crc_t was removed in https://github.com/zlib-ng/zlib-ng/commit/4db4cfdb5badc8860f7410732b12c45216d709b3. Do we need to bring it back for zlib.h ZLIB_COMAT?

mtl1979 replied:
All optimized crc32 functions should use only `uint32_t` as CRC32 variable type, the conversion for ZLIB_COMPAT API is done in `crc32_z()` and `crc32()` inside crc32.c. All length parameters should always be `size_t`. Conversion from `unsigned int` is also done in crc32.c.

nmoinvaz replied:
get_crc_table may have produced unsigned long return type in zlib.

mscastanho replied:
If we change the test to just call `crc32_z` as @mtl1979 suggested, these typedefs become irrelevant, as most types are the same between compat and zlib-ng. But the `crc` still has different types on each.

mtl1979 replied:
If the test is statically linked to zlib-ng, it can access functable directly, and such there will be no differences... For dynamic linking, simple cast when calling `crc32_z` is enough. I pointed out to how it is done in crc32.c, because essentially what is needed is same backwards. It would be a lot easier, if there would be "undocumented API" to access functable when building against shared version of zlib-ng.

mtl1979 commented on test/crc32_test.c:
```diff
@@ -0,0 +1,208 @@
+} crc32_test;
+
+void test_crc32(crc32_type crc, buf_type *buf, len_type len, crc32_type chk, uint32_t line) {
+    crc32_type res = PREFIX(crc32)(crc, buf, len);
```
Should use `crc32_z` as the old function without suffix is deprecated and simply wrapper to new function.

mscastanho replied:
Ok, replaced by `crc32_z`.

mtl1979 commented on test/crc32_test.c:
```diff
@@ -0,0 +1,208 @@
+#ifdef ZLIB_COMPAT
+#  include "zlib.h"
+typedef unsigned long crc32_type;
+typedef unsigned char buf_type;
```
Should be `const unsigned char`... Assigning string literals to `unsigned char` is deprecated.

mtl1979 commented on test/crc32_test.c:
```diff
@@ -0,0 +1,208 @@
+#else
+#  include "zlib-ng.h"
+typedef uint32_t crc32_type;
+typedef uint8_t buf_type;
```
Should be `const uint8_t`.  Assigning string literals to non-const types is deprecated.

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/clang_workaround.h:
```diff
@@ -0,0 +1,98 @@
+/* Helper functions to work around issues with clang builtins
```
There is a `fallback_builtins.h` in the main directory. I think it should be moved to arch/x86/fallback_builtins.h. And then this file should be renamed arch/power/fallback_builtins.h.

mscastanho replied:
I renamed `clang_workarounds.h` to `fallback_builtins.h`. I haven't moved the x86 one though.

nmoinvaz commented on arch/power/clang_workaround.h:
```diff
@@ -0,0 +1,98 @@
+__vector unsigned long long  __builtin_pack_vector (unsigned long __a,
+						    unsigned long __b)
+{
+	#if defined(__BIG_ENDIAN__)
```
No indent is necessary for preprocess statements in this source file.

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/clang_workaround.h:
```diff
@@ -0,0 +1,98 @@
+static inline
+__vector unsigned long long __attribute__((overloadable))
+vec_ld(int __a, const __vector unsigned long long* __b)
+{
```
Brackets should be on same line as function definition for this file.

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/clang_workaround.h:
```diff
@@ -0,0 +1,98 @@
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#ifndef CLANG_WORKAROUNDS_H
```
Use similar to fallback_builtins.h `POWER_BUILTINS_H`?

mscastanho replied:
Done.

nmoinvaz commented on arch/power/clang_workaround.h:
```diff
@@ -0,0 +1,98 @@
+ */
+static inline
+__vector unsigned long long  __builtin_pack_vector (unsigned long __a,
+						    unsigned long __b)
```
Second argument in a few of these functions don't align. And there is space after function names.

mscastanho replied:
I changed so all are in the same line now.

nmoinvaz commented on arch/power/crc32_constants.h:
```diff
@@ -0,0 +1,1204 @@
+
+#ifndef __ASSEMBLER__
+#ifdef CRC_TABLE
+static const unsigned int crc_table[] = {
```
This table appears to be the same as in /crc32_tbl.h.

mscastanho replied:
You're right! It's the same as `crc_table[0]`. I'll change the code to use that.

mscastanho replied:
The table has been removed in favor of reusing `crc_table[0]` from `crc32_tbl.h`.

nmoinvaz commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,667 @@
+#if defined(__LITTLE_ENDIAN__)
+/* Byte reverse permute constant LE. */
+static const __vector unsigned long long vperm_const
+	__attribute__ ((aligned(16))) = { 0x08090A0B0C0D0E0FUL,
```
* All of the `__attribute__ aligned` statements should use zutil.h's `ALIGNED_` macro.
* This source code file suffers from some of the same style inconsistencies as the other.

mscastanho replied:
I replaced all usages of the attribute by the macro and made styling changes to the entire file. Should be fine now.

nmoinvaz commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,667 @@
+			/* We reduce our final 128 bytes in a separate step */
+			chunks = (block_size/128)-1;
+
+		    vconst1 = vec_ld(offset, vcrc_const);
```
Weird alignment. All the source files should use spaces instead of tabs. :-)

mscastanho replied:
Fixed.

nmoinvaz reviewed (commented):
I have made some initial comments mainly on style. Thank you for working to get this relicensed for zlib! I think once this PR is merged the library will be better for it.

mscastanho commented:
Applied many styling changes suggested. I also realized `crc32_constants.h` could be simplified, as the original file was more generic and had some `ifdefs` to change the code for different scenarios. But for this case we were always using it the same way, so I removed the `ifdefs` and macros (`REFLECT`, `CRC_XOR`, `POWER8_INTRINSICS`) and left the corresponding code.

mtl1979 commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,87 @@
+/*
+ * GCC __builtin_pack_vector_int128 returns a vector __int128_t but Clang
+ * does not recognize this type. On GCC this builtin is translated to a
+ * xxpermdi instruction that only moves the registers __a, __b instead generates
```
"instead generates" sounds weird, maybe "instead of generating" or something like that.

mscastanho replied:
Agreed. Fixed.

mtl1979 commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,87 @@
+
+/*
+ * Clang 7 changed the behavior of vec_xxpermdi in order to provide the same
+ * behavior of GCC. That means code adapted to Clang >= 7 does not work on
```
"same behavior as"
"adapted for"

mscastanho replied:
Fixed.

mtl1979 commented on test/crc32_test.c:
```diff
@@ -0,0 +1,202 @@
+  {__LINE__, 0x508fae6a, (unsigned char *)"0l1tw7GOcem06Ddu7yn4", 20, 0xb52b38bc},
+  {__LINE__, 0xe5adaf4f, (unsigned char *)"MCr47CjPIn9R1IvE1Tm5", 20, 0xf83b8178},
+  {__LINE__, 0x67136a40, (unsigned char *)"UcixbzPKTIv0SvILHVdO", 20, 0xc5213070},
+  {__LINE__, 0xb00c4a10, (unsigned char *)"dGnAyAhRQDsWw0ESou24", 20, 0xbc7648b0},
+  {__LINE__, 0x2e0c84b5, (unsigned char *)"di0nvmY9UYMYDh0r45XT", 20, 0xd8123a72},
+  {__LINE__, 0x81238d44, (unsigned char *)"2XKDwHfAhFsV0RhbqtvH", 20, 0xd5ac5620},
+  {__LINE__, 0xf853aa92, (unsigned char *)"ZhrANFIiIvRnqClIVyeD", 20, 0xceae099d},
+  {__LINE__, 0x5a692325, (unsigned char *)"v7Q9ehzioTOVeDIZioT1", 20, 0xb07d2b24},
+  {__LINE__, 0x3275b9f, (unsigned char *)"Yod5hEeKcYqyhfXbhxj2", 20, 0x24ce91df},
+  {__LINE__, 0x38371feb, (unsigned char *)"GehSWY2ay4uUKhehXYb0", 20, 0x707b3b30},
+  {__LINE__, 0xafc8bf62, (unsigned char *)"kwytJmq6UqpflV8Y8GoE", 20, 0x16abc6a9},
+  {__LINE__, 0x9b07db73, (unsigned char *)"70684206568419061514", 20, 0xae1fb7b7},
+  {__LINE__, 0xe75b214, (unsigned char *)"42015093765128581010", 20, 0xd4eecd2d},
+  {__LINE__, 0x72d0fe6f, (unsigned char *)"88214814356148806939", 20, 0x4660ec7},
+  {__LINE__, 0xf857a4b1, (unsigned char *)"43472694284527343838", 20, 0xfd8afdf7},
+  {__LINE__, 0x54b8e14, (unsigned char *)"49769333513942933689", 20, 0xc6d1b5f2},
+  {__LINE__, 0xd6aa5616, (unsigned char *)"54979784887993251199", 20, 0x32476461},
+  {__LINE__, 0x11e63098, (unsigned char *)"58360544869206793220", 20, 0xd917cf1a},
+  {__LINE__, 0xbe92385, (unsigned char *)"27347953487840714234", 20, 0x4ad14a12},
+  {__LINE__, 0x49511de0, (unsigned char *)"07650690295365319082", 20, 0xe37b5c6c},
+  {__LINE__, 0x3db13bc1, (unsigned char *)"42655507906821911703", 20, 0x7cc497f1},
+  {__LINE__, 0xbb899bea, (unsigned char *)"29977409200786225655", 20, 0x99781bb2},
+  {__LINE__, 0xf6cd9436, (unsigned char *)"85181542907229116674", 20, 0x132256a1},
+  {__LINE__, 0x9109e6c3, (unsigned char *)"87963594337989416799", 20, 0xbfdb2c83},
+  {__LINE__, 0x75770fc, (unsigned char *)"21395988329504168551", 20, 0x8d9d1e81},
+  {__LINE__, 0x69b1d19b, (unsigned char *)"51991013580943379423", 20, 0x7b6d4404},
+  {__LINE__, 0xc6132975, (unsigned char *)"*]+@!);({_$;}[_},?{?;(_?,=-][@", 30, 0x8619f010},
+  {__LINE__, 0xd58cb00c, (unsigned char *)"_@:_).&(#.[:[{[:)$++-($_;@[)}+", 30, 0x15746ac3},
+  {__LINE__, 0xb63b8caa, (unsigned char *)"&[!,[$_==}+.]@!;*(+},[;:)$;)-@", 30, 0xaccf812f},
+  {__LINE__, 0x8a45a2b8, (unsigned char *)"]{.[.+?+[[=;[?}_#&;[=)__$$:+=_", 30, 0x78af45de},
+  {__LINE__, 0xcbe95b78, (unsigned char *)"-%.)=/[@].:.(:,()$;=%@-$?]{%+%", 30, 0x25b06b59},
+  {__LINE__, 0x4ef8a54b, (unsigned char *)"+]#$(@&.=:,*];/.!]%/{:){:@(;)$", 30, 0x4ba0d08f},
+  {__LINE__, 0x76ad267a, (unsigned char *)")-._.:?[&:.=+}(*$/=!.${;(=$@!}", 30, 0xe26b6aac},
+  {__LINE__, 0x569e613c, (unsigned char *)":(_*&%/[[}+,?#$&*+#[([*-/#;%(]", 30, 0x7e2b0a66},
+  {__LINE__, 0x36aa61da, (unsigned char *)"{[#-;:$/{)(+[}#]/{&!%(@)%:@-$:", 30, 0xb3430dc7},
+  {__LINE__, 0xf67222df, (unsigned char *)"_{$*,}(&,@.)):=!/%(&(,,-?$}}}!", 30, 0x626c17a},
+  {__LINE__, 0x74b34fd3, (unsigned char *)"e$98KNzqaV)Y:2X?]77].{gKRD4G5{mHZk,Z)SpU%L3FSgv!Wb8MLAFdi{+fp)c,@8m6v)yXg@]HBDFk?.4&}g5_udE*JHCiH=aL", 100, 0xccf98060},
+  {__LINE__, 0x351fd770, (unsigned char *)"r*Fd}ef+5RJQ;+W=4jTR9)R*p!B;]Ed7tkrLi;88U7g@3v!5pk2X6D)vt,.@N8c]@yyEcKi[vwUu@.Ppm@C6%Mv*3Nw}Y,58_aH)", 100, 0xd8b95312},
+  {__LINE__, 0xc45aef77, (unsigned char *)"h{bcmdC+a;t+Cf{6Y_dFq-{X4Yu&7uNfVDh?q&_u.UWJU],-GiH7ADzb7-V.Q%4=+v!$L9W+T=bP]$_:]Vyg}A.ygD.r;h-D]m%&", 100, 0xbb1c9912},
+  {__LINE__, 0xc45aef77, (unsigned char *)"h{bcmdC+a;t+Cf{6Y_dFq-{X4Yu&7uNfVDh?q&_u.UWJU],-GiH7ADzb7-V.Q%4=+v!$L9W+T=bP]$_:]Vyg}A.ygD.r;h-D]m%&"
```
Missing "const"...

mscastanho replied:
Fixed.

mtl1979 commented on test/crc32_test.c:
```diff
@@ -0,0 +1,202 @@
+void test_crc32(crc32_type crc, const unsigned char *buf, size_t len, uint32_t chk, uint32_t line) {
+    uint32_t res = PREFIX(crc32_z)((uint32_t)crc, buf, len);
+    if (res != chk) {
+        fprintf(stderr, "FAIL [%d]: crc32 returned 0x%08X expected 0x%08X\n",
+                line, res, chk);
```
This line doesn't seem longer than line 29.

mscastanho replied:
Fixed.

mscastanho commented:
Rebased against latest develop branch.

mscastanho commented:
@mtl1979 @nmoinvaz Do you have any other suggestions?

mscastanho commented:
Rebase.

mtl1979 commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,87 @@
+#define __builtin_crypto_vpmsumd __builtin_crypto_vpmsumb
+#endif
+
+static inline
+__vector unsigned long long __attribute__((overloadable))
```
Fits in single line?

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/crc32_constants.h:
```diff
@@ -0,0 +1,1122 @@
+
+/* Reduce 262144 kbits to 1024 bits */
+static const __vector unsigned long long vcrc_const[255] ALIGNED_(16) = {
+#ifdef __LITTLE_ENDIAN__
```
For endian detection use `BYTE_ORDER == LITTLE_ENDIAN` (see zendian.h)

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,87 @@
+ * Clang has vec_xxpermdi intrinsics. It was implemented in 4.0.0.
+ */
+static inline __vector unsigned long long __builtin_pack_vector(unsigned long __a, unsigned long __b) {
+#if defined(__BIG_ENDIAN__)
```
For endian detection use `BYTE_ORDER == BIG_ENDIAN` (see zendian.h)

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,87 @@
+    return __v[__o];
+}
+
+#if defined(__BIG_ENDIAN__)
```
For endian detection use `BYTE_ORDER == BIG_ENDIAN` (see zendian.h)

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,87 @@
+}
+
+static inline unsigned long __builtin_unpack_vector_1(__vector unsigned long long __v) {
+#if defined(__BIG_ENDIAN__)
```
For endian detection use `BYTE_ORDER == BIG_ENDIAN` (see zendian.h)

mscastanho replied:
Fixed.

nmoinvaz commented on test/crc32_test.c:
```diff
@@ -0,0 +1,201 @@
+#include "zbuild.h"
+#ifdef ZLIB_COMPAT
+#  include "zlib.h"
+typedef unsigned long crc32_type;
```
I'm not sure this is entirely correct. It should be `typedef z_crc_t crc32_type` correct? Because depending on the version of zlib and config it can be either unsigned int or unsigned long. Some older versions of zlib don't have `z_crc_t` either.

nmoinvaz replied:
Maybe for this file it is better to just use `unsigned long` every where and it will be cast to `uint32_t` in `crc32_z` call..

mscastanho replied:
Fair enough. I changed the code to use `unsigned long` everywhere.

nmoinvaz reviewed (commented):
Ok a few more comments.

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,595 @@
+#define VMX_ALIGN	16
+#define VMX_ALIGN_MASK	(VMX_ALIGN-1)
+
+static unsigned int crc32_align(unsigned int crc, const unsigned char *p, unsigned long len)
```
`{` should be on same line.

mscastanho replied:
Fixed

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,595 @@
+static unsigned int ALIGNED_(32) __crc32_vpmsum(unsigned int crc, const void* p, unsigned long len);
+
+Z_INTERNAL uint32_t crc32_power8(uint32_t crc, const unsigned char *p, uint64_t _len)
+{
```
`{` should be on same line.

mscastanho replied:
Fixed.

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,593 @@
+#if defined (__clang__)
+#include "fallback_builtins.h"
+#else
+#define __builtin_pack_vector(a, b)  __builtin_pack_vector_int128 ((a), (b))
```
This seems unnecessary... parameters are passed in same order without any additional cast.

mscastanho replied:
This was needed because an alternative `__builtin_pack_vector` was defined in `fallback_builtins` for clang, so we needed a definition for the GCC case. But clang now has `__builtin_pack_vector_int128`, so we can use that directly instead, so this was removed.

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,593 @@
+
+                    v0 = vec_xor(v0, va0);
+                    va0 = __builtin_crypto_vpmsumd ((__vector unsigned long
+                                                     long)vdata0, (__vector unsigned long long)vconst2);
```
Whole cast should be on same line, second parameter on own line.

mscastanho replied:
@mtl1979 this has already been fixed on the last push.

mtl1979 replied:
@mscastanho That's just GitHub acting up... It reposts all comments when finishing review...

mtl1979 commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,88 @@
+ *
+ * Clang has vec_xxpermdi intrinsics. It was implemented in 4.0.0.
+ */
+static inline __vector unsigned long long __builtin_pack_vector(unsigned long __a, unsigned long __b) {
```
Clang does support vector containing single 128-bit unsigned integer using type:
`vector unsigned __int128`

This is one of the functions that benefits from features added in Power10. Basically Power10 can load and merge using one instruction, instead of two loads + vector permute using Power8 and Power9.

mscastanho replied:
This snippet is not needed anymore with the latest code.

mtl1979 commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,88 @@
+
+#else
+
+static inline unsigned long __builtin_unpack_vector_0(__vector unsigned long long __v) {
+#if BYTE_ORDER == BIG_ENDIAN
+    return vec_xxpermdi(__v, __v, 0x0)[0];
+#else
+    return vec_xxpermdi(__v, __v, 0x3)[0];
+#endif
+}
+
+static inline unsigned long __builtin_unpack_vector_1(__vector unsigned long long __v) {
+#if BYTE_ORDER == BIG_ENDIAN
+    return vec_xxpermdi(__v, __v, 0x3)[0];
+#else
+    return vec_xxpermdi(__v, __v, 0x0)[0];
+#endif
+}
+#endif /* vec_xxpermdi */
```
I think this is just adding clutter... Using vec_extract() or just array notation would work on all compiler versions... Power9 and newer even have own assembler instruction for vec_extract().

mscastanho replied:
I replaced the usages of `__builtin_unpack_vector_*` with array notation. So this has been removed.

mtl1979 commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,88 @@
+ * Clang has vec_xxpermdi intrinsics. It was implemented in 4.0.0.
+ */
+static inline __vector unsigned long long __builtin_pack_vector(unsigned long __a, unsigned long __b) {
+#if BYTE_ORDER == BIG_ENDIAN
+    __vector unsigned long long __v = {__a, __b};
+#else
+    __vector unsigned long long __v = {__b, __a};
+#endif
```
For little-endian, just add `__v = vec_reve(__v);`

This avoids enclosing the big-endian part in preprocessor condition. The generated code is better in debug mode, but same in release mode.

mscastanho replied:
Not needed anymore.

Dead2 commented:
Sorry, needs another little rebase.

mscastanho commented:
@Dead2 rebase done.

Dead2 commented:
Just needs some approvals now 😉

mtl1979 commented:
@Dead2 The patch is quite big, so it takes a long time to review... I might know something about PowerPC/Power intrinsics and how they map to assembler instructions, but other people might be better at looking style inconsistencies...

nmoinvaz commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,593 @@
+
+    unsigned long len = (unsigned long) _len;
+
+    if (p == (const unsigned char *) 0x0) return 0;
```
Separate line for `return 0`;

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,593 @@
+                offset += 16;
+
+                v0 = vec_xor(v0, va0);
+                va0 = __builtin_crypto_vpmsumd ((__vector unsigned long
```
Random functions with space after function name above and below this line.

mscastanho replied:
Ok, I think I fixed all of those now.

nmoinvaz commented on arch/power/fallback_builtins.h:
```diff
@@ -0,0 +1,88 @@
+}
+
+#if BYTE_ORDER == BIG_ENDIAN
+#define __builtin_unpack_vector_0(a) __builtin_unpack_vector ((a), 0)
```
Extra space after function name here too.

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,593 @@
+        offset += 128;
+
+        v0 = (__vector unsigned long long)__builtin_crypto_vpmsumw (
+            (__vector unsigned int)vdata0,(__vector unsigned int)v0);
```
No space between function arguments.

mscastanho replied:
Fixed.

nmoinvaz commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,593 @@
+             * left 32 bits so it occupies the least significant bits in the
+             * bit reflected domain.
+             */
+            v0 = (__vector unsigned long long)vec_sld((__vector unsigned char)v0,
```
@mscastanho what do you think about using typedef?
```
_vector unsigned long long = vector_uint64_t?
_vector unsigned char = vector_uint8_t?
```
Or just using fixed types if possible?

Anyways, just an idea because I noticed there is a lot of casting with long types.

mscastanho replied:
@nmoinvaz I don't think there are vector types with "fixed type names" available for Power (something like `vector uint64_t`). But the types available do have fixed sizes (e.g. `vector unsigned long long` is a vector with 2 doublewords).

I think I'm neutral regarding using typedefs. It would certainly save some chars and make the code neater, but would add an extra indirection for the "reader" to find out which underlying type is actually being used.

If you think using typedefs would be better, I can make the change.

nmoinvaz reviewed (commented):
I only have style changes as I don't have the hardware. It is good that there are crc32 tests to prove it out.

mscastanho commented:
@nmoinvaz I'm so sorry. I completely missed your last comments 🤦🏼 

I  think I addressed all of them, and also rebased against current develop to fix some merge conflicts.

nmoinvaz commented:
LGTM.

mscastanho commented:
Gentle ping =)

mtl1979 commented on Makefile.in:
```diff
@@ -288,6 +291,10 @@ endif
 
 adler32_test$(EXE): adler32_test.o $(OBJG) $(STATICLIB)
 	$(CC) $(CFLAGS) $(LDFLAGS) -o $@ adler32_test.o $(OBJG) $(TEST_LIBS) $(LDSHAREDLIBC)
+
+crc32_test$(EXE): crc32_test.o $(STATICLIB)
+	$(CC) $(LDFLAGS) -o $@ crc32_test.o $(TEST_LIBS) $(LDSHAREDLIBC)
+
```
This removes executing strip on `adler32_test$(EXE)`!

mscastanho replied:
Fixed.

mtl1979 reviewed (changes requested):
Needs some cleanup.

mtl1979 commented on Makefile.in:
```diff
@@ -292,6 +295,13 @@ ifneq ($(STRIP),)
 	$(STRIP) $@
 endif
 
+
```
Extra empty line ;)

mscastanho replied:
Fixed.

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,590 @@
+}
+
+/* When we have a load-store in a single-dispatch group and address overlap
+ * such that foward is not allowed (load-hit-store) the group must be flushed.
```
"foward" is most likely typo.

mscastanho replied:
Fixed

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,590 @@
+        vdata0 = vec_ld(0, (__vector unsigned long long*) p);
+        VEC_PERM(vdata0, vdata0, vconst1, vperm_const);
+
+        /* xor initial value*/
```
Missing space before `*/`.

mscastanho replied:
Fixed

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,590 @@
+                    p = (char *)p + 128;
+                }
+
+                /* First cool down*/
```
Missing space before `*/`.

mscastanho replied:
Fixed

mtl1979 reviewed (changes requested):
Just small style fixes ;)

Dead2 commented:
This needs a minor rebase now, sorry 😉

mscastanho commented:
@Dead2 Done =)

Dead2 reviewed (approved):
LGTM

mtl1979 commented on arch/power/crc32_power8.c:
```diff
@@ -0,0 +1,590 @@
+                p = (char *)p + 128;
+
+                /*
+                 * main loop. We modulo schedule it such that it takes three
```
This line is hard to read. Possibly missing punctuation.

mscastanho replied:
This comment was inherited from previous iterations of this code. I replaced it with simpler high-level comment.

mtl1979 reviewed (changes requested):
There is still quite a lot of casts that make the code hard to read, but I don't think keeping them blocks merging as trying to avoid casts by using inline utility functions makes debug build larger.

mtl1979 reviewed (approved):
LGTM.
